## Libraries

In [2]:
import pandas as pd, matplotlib.pyplot as plt, numpy as np

# Project Details

To predict the probability of bookings being cancelled

In this notebook it's about data exploration and data analysis

purpose:

Briefly go over the dataset > brief understanding of the dataset itself

investigate features > whether they are useful or not, if they are data leakage, or obvious identifiers 

identify the target variable 

find patterns that are linked to the probability of bookings being cancelled 




## Dataset

In [3]:
df = pd.read_csv('../data/hotel_bookings.csv')
df

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

## Dataset Overview

In this section, we examine the structure of the dataset to build a foundational understanding before deeper analysis. Specifically, we review: 

* the number of observations and features
* data types of each feature
* presence of missing values
* duplicate records

#### Observations/features

In [4]:
df.shape

(119390, 32)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

#### Summary

* 32 features and 119,390 records

* 20 numerical features, and 12 categorical features

* Next step is to analyse missing values both in numerical and categorical 


#### Missing Values

In [6]:
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [7]:
company = 112593/df.shape[0]
agent = 16340/df.shape[0]
country = 488/df.shape[0]

company, agent, country

(0.943068933746545, 0.13686238378423654, 0.004087444509590418)

By doing the `isnull` method we get a brief overview of empty values. 

Features that have missing values above:

Company, Agent: An overwhelming amount of missing values, therefore, requires investigation into it

Country/ children: Minute amount of missing values


In [8]:
### Build a function that will identify objects (categorical data types), and iterate over them with the unique function to see if there are hidden missing values as text

def display_categorical_data_type_as_unqiue(df):
    # retrieve the names under object data types
    columns = df.columns
    for x in columns:
        if df[x].dtypes == 'object':
            print(x)
            print(df[x].unique())

display_categorical_data_type_as_unqiue(df)

hotel
['Resort Hotel' 'City Hotel']
arrival_date_month
['July' 'August' 'September' 'October' 'November' 'December' 'January'
 'February' 'March' 'April' 'May' 'June']
meal
['BB' 'FB' 'HB' 'SC' 'Undefined']
country
['PRT' 'GBR' 'USA' 'ESP' 'IRL' 'FRA' nan 'ROU' 'NOR' 'OMN' 'ARG' 'POL'
 'DEU' 'BEL' 'CHE' 'CN' 'GRC' 'ITA' 'NLD' 'DNK' 'RUS' 'SWE' 'AUS' 'EST'
 'CZE' 'BRA' 'FIN' 'MOZ' 'BWA' 'LUX' 'SVN' 'ALB' 'IND' 'CHN' 'MEX' 'MAR'
 'UKR' 'SMR' 'LVA' 'PRI' 'SRB' 'CHL' 'AUT' 'BLR' 'LTU' 'TUR' 'ZAF' 'AGO'
 'ISR' 'CYM' 'ZMB' 'CPV' 'ZWE' 'DZA' 'KOR' 'CRI' 'HUN' 'ARE' 'TUN' 'JAM'
 'HRV' 'HKG' 'IRN' 'GEO' 'AND' 'GIB' 'URY' 'JEY' 'CAF' 'CYP' 'COL' 'GGY'
 'KWT' 'NGA' 'MDV' 'VEN' 'SVK' 'FJI' 'KAZ' 'PAK' 'IDN' 'LBN' 'PHL' 'SEN'
 'SYC' 'AZE' 'BHR' 'NZL' 'THA' 'DOM' 'MKD' 'MYS' 'ARM' 'JPN' 'LKA' 'CUB'
 'CMR' 'BIH' 'MUS' 'COM' 'SUR' 'UGA' 'BGR' 'CIV' 'JOR' 'SYR' 'SGP' 'BDI'
 'SAU' 'VNM' 'PLW' 'QAT' 'EGY' 'PER' 'MLT' 'MWI' 'ECU' 'MDG' 'ISL' 'UZB'
 'NPL' 'BHS' 'MAC' 'TGO' 'TWN' 'DJI' 'STP' 'KNA' 'ETH' 'IR

The following categorical features have missing values:
* meal has undefined 
* market_segment has undefined
* distribution_channel has undefined

The next step here would be to calculate the total amount of missing values in the above listed features


In [9]:
def custom_sum_missing_values(df,feature):
    undefined = 0

    for x in df[feature]:
        if x == 'Undefined':
            undefined += 1
    percentage = undefined/len(df)        
    print(f"{feature} undefined values: {undefined} %: {percentage} %")
            

In [10]:
custom_sum_missing_values(df,'meal')
custom_sum_missing_values(df,'market_segment')
custom_sum_missing_values(df, 'distribution_channel')

meal undefined values: 1169 %: 0.009791439819080325 %
market_segment undefined values: 2 %: 1.6751821760616466e-05 %
distribution_channel undefined values: 5 %: 4.1879554401541165e-05 %


After going through the missing values from categorical features. It seems like they are all minute amount, therefore deleteing these records will not be an issue.

Next, I will examine missing values that are abundant in company and agent. Both these features are numerical representations  

In [11]:
df.distribution_channel.value_counts()

distribution_channel
TA/TO        97870
Direct       14645
Corporate     6677
GDS            193
Undefined        5
Name: count, dtype: int64

In [12]:
df.market_segment.value_counts()

market_segment
Online TA        56477
Offline TA/TO    24219
Groups           19811
Direct           12606
Corporate         5295
Complementary      743
Aviation           237
Undefined            2
Name: count, dtype: int64

In [13]:
df.meal.value_counts()

meal
BB           92310
HB           14463
SC           10650
Undefined     1169
FB             798
Name: count, dtype: int64

In [14]:
pd.crosstab(df['meal'],df['is_canceled'])

is_canceled      0      1
meal                     
BB           57800  34510
FB             320    478
HB            9479   4984
SC            6684   3966
Undefined      883    286

In [15]:
pd.crosstab(df['meal'],df['is_canceled'],normalize='index')

is_canceled         0         1
meal                           
BB           0.626151  0.373849
FB           0.401003  0.598997
HB           0.655397  0.344603
SC           0.627606  0.372394
Undefined    0.755346  0.244654

In [16]:
pd.crosstab(df['meal'],df['is_canceled'],normalize='columns')

is_canceled         0         1
meal                           
BB           0.768965  0.780346
FB           0.004257  0.010809
HB           0.126108  0.112699
SC           0.088923  0.089680
Undefined    0.011747  0.006467

standard of error: ${SE = \sqrt{\frac{p(1-p)}{n}}}$
signal-to-noise-ratio: ${signal-to-noise-ratio = \frac{difference}{SE}}$

The feature meal required further examination. This is because of a hypthoesis i concluded: I assume undefined is similar to self-catering infact it should be identical. This is because I believe when customers leave it blank that means they bring their own meal or not have any meal selected. This references self-catering. The above `pd.crosstab` illustrates if this hypothesis is valid or not. Roughly, the cancellation rate are 37% and 24% for self-catering and undefined, respectively. This illustrates that self-catering and undefined follow a different trend. Therefore, deleting the category undefined would delete a signal that could be beneficial for modelling. If they were similar, then merging would have been an option, but that's not the case here. One important nuance to take into consideration would be the sample size. The sample size determines how reliable is one probability is. In this case, Undefined is a sample size of 1169, a proportion of 0.009% of the meal feature. However, after doing `.value_counts()` it becomes evident that there's a small category in the meal feature, FB. Therefore I conclude, there's not enough reasons or evidence statiscally to remove undefined, instead I have gathered reasons to keep it, one of them being it brings a signal, it's not a noise. This will help the model to build more sophisticated patterns. I will alter the name of undefined to 'no meals', this would infer no meals are selected or that it's left blank from the customer.

The feature meal required further investigation due to the presence of the category Undefined. A hypothesis was formed that Undefined may represent customers who did not select a meal plan and therefore could be behaviourally similar to Self-Catering (SC).

To test this hypothesis, a conditional probability analysis was conducted using pd.crosstab, examining the cancellation rate across meal categories. The results showed that SC has a cancellation rate of approximately 37%, whereas Undefined has a significantly lower cancellation rate of approximately 24%.

This substantial difference suggests that Undefined does not follow the same behavioural pattern as SC. Therefore, merging these categories would result in the loss of a distinct predictive signal.

Although Undefined represents approximately 1% of the dataset (n = 1169), this sample size is sufficiently large to produce a stable probability estimate. Additionally, another category (FB) has an even smaller sample size, indicating that Undefined is not uniquely small within this feature.

Based on this analysis, there is insufficient statistical justification to remove or merge the Undefined category. Instead, it will be retained and renamed to No_Meal to improve interpretability while preserving its predictive value.

In [17]:
df.company.isnull()

0         True
1         True
2         True
3         True
4         True
          ... 
119385    True
119386    True
119387    True
119388    True
119389    True
Name: company, Length: 119390, dtype: bool

After doing research regarding company it turns out, company is actually the id of the company or cooperate that made the booking for this individual. Now, since most of the booking that's made are actually invdividuals, so regular customers. Which explains why there's so many missing values. Now, I could potentially recode the null values as non-cooperate/non-company, but this would later lead to sparseness, making it harder when it comes to the model for the models. Therefore, I have decided to drop the company feature itself. 

In [18]:
df.agent

0           NaN
1           NaN
2           NaN
3         304.0
4         240.0
          ...  
119385    394.0
119386      9.0
119387      9.0
119388     89.0
119389      9.0
Name: agent, Length: 119390, dtype: float64

In [19]:
from collections import Counter
Counter(df['distribution_channel'] == 'Direct')

Counter({False: 104745, True: 14645})

In [20]:
Counter(df.agent == 0)

Counter({False: 119390})

After further research regarding agent, it turns out, agent refers to the ID of the travel agency that made the booking. So the missing values actually infers direct booking. Since, it's not a large quantative amount compared to company, I will note remove the records or the feature itself instead I will replace the empty value as 0 representing no agency involved.

In [21]:
df.duplicated().sum()

np.int64(31994)

In [22]:
df[df.duplicated()]

hotel  is_canceled  lead_time  arrival_date_year  \
5       Resort Hotel            0         14               2015   
22      Resort Hotel            0         72               2015   
43      Resort Hotel            0         70               2015   
138     Resort Hotel            1          5               2015   
200     Resort Hotel            0          0               2015   
...              ...          ...        ...                ...   
119349    City Hotel            0        186               2017   
119352    City Hotel            0         63               2017   
119353    City Hotel            0         63               2017   
119354    City Hotel            0         63               2017   
119373    City Hotel            0        175               2017   

       arrival_date_month  arrival_date_week_number  \
5                    July                        27   
22                   July                        27   
43                   July                        27   
138                  July                        28   
200                  July                        28   
...                   ...                       ...   
119349             August                        35   
119352             August                        35   
119353             August                        35   
119354             August                        35   
119373             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
5                               1                        0   
22                              1                        2   
43                              2                        2   
138                             5                        1   
200                             7                        0   
...                           ...                      ...   
119349                         31                        0   
119352                         31                        0   
119353                         31                        0   
119354                         31                        0   
119373                         31                        1   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
5                          2       2  ...    No Deposit  240.0     NaN   
22                         4       2  ...    No Deposit  250.0     NaN   
43                         3       2  ...    No Deposit  250.0     NaN   
138                        0       2  ...    No Deposit  240.0     NaN   
200                        1       1  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119349                     3       2  ...    No Deposit    9.0     NaN   
119352                     3       3  ...    No Deposit    9.0     NaN   
119353                     3       3  ...    No Deposit    9.0     NaN   
119354                     3       3  ...    No Deposit    9.0     NaN   
119373                     3       1  ...    No Deposit   42.0     NaN   

       days_in_waiting_list    customer_type     adr  \
5                         0        Transient   98.00   
22                        0        Transient   84.67   
43                        0        Transient  137.00   
138                       0        Transient   97.00   
200                       0        Transient  109.80   
...                     ...              ...     ...   
119349                    0        Transient  126.00   
119352                    0  Transient-Party  195.33   
119353                    0  Transient-Party  195.33   
119354                    0  Transient-Party  195.33   
119373                    0        Transient   82.35   

        required_car_parking_spaces  total_of_special_requests  \
5                                 0                          1   
22                                0                          1   
43                                0         

It seems like there 31,994 duplicate records. Upon further examination there's no unique identifier that would uniquely identify that each record is made by the same person/individual. This means that, many individuals could've booked the same book details but two different individuals. Therefore, no actions will take place here.

## Handling missing data 

* Company: Erase the entire feature
* Agent: Replace empty values with a more useful insight value 0
* Country/Children: Minute amount, therefore erase the records
* Distribution Channel/ Market Segement: Erase the records that are missing
* meal replace undefined with no meal

In [111]:
len(df.columns)

32

In [112]:
# dropping the feature company
df_reduced = df.drop(columns=['company'])
len(df_reduced.columns)

31

In [113]:
len(df_reduced)

119390

In [114]:
# dropping the country, and column rows that are empty
df_reduced = df_reduced.dropna(subset=['country','children'])
len(df_reduced)

118898

In [115]:
# drop the rows from the feature: distribution channel and market segement, where there are undefined 

COLUMNS_TO_CLEAN = ["market_segment", "distribution_channel"]
VALUES_TO_REMOVE = ["Undefined"]

mask = df_reduced[COLUMNS_TO_CLEAN].isin(VALUES_TO_REMOVE).any(axis=1)
df_reduced = df_reduced.loc[~mask].copy()

In [116]:
len(df_reduced), len(df_reduced.columns)

(118897, 31)

In [117]:
df_reduced.agent.isnull().sum()

np.int64(16003)

In [118]:
# Agent we want to replace NaN with 0
df_reduced['agent'] = df_reduced['agent'].fillna(0)
df_reduced.agent.isnull().sum()

np.int64(0)

In [119]:
(df_reduced['meal'] == 'Undefined').sum()

np.int64(1165)

In [120]:
# meals that are undefined, replace with no meal

df_reduced['meal'] = df_reduced['meal'].replace("Undefined","No meal")
(df_reduced['meal'] == 'Undefined').sum()
(df_reduced['meal'] == 'No meal').sum()

np.int64(1165)

In [121]:
df_reduced

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  booking_changes  deposit_type  \
0                          0       2  ...                3    No Deposit   
1                          0       2  ...                4    No Deposit   
2                          1       1  ...                0    No Deposit   
3                          1       1  ...                0    No Deposit   
4                          2       2  ...                0    No Deposit   
...                      ...     ...  ...              ...           ...   
119385                     5       2  ...                0    No Deposit   
119386                     5       3  ...                0    No Deposit   
119387                     5       2  ...                0    No Deposit   
119388                     5       2  ...                0    No Deposit   
119389                     7       2  ...                0    No Deposit   

        agent days_in_waiting_list customer_type     adr  \
0         0.0                    0     Transient    0.00   
1         0.0                    0     Transient    0.00   
2         0.0                    0     Transient   75.00   
3       304.0                    0     Transient   75.00   
4       240.0                    0     Transient   98.00   
...       ...                  ...           ...     ...   
119385  394.0                    0     Transient   96.14   
119386    9.0                    0     Transient  225.43   
119387    9.0                    0     Transient  157.71   
119388   89.0                    0     Transient  104.40   
119389    9.0                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0   

In [122]:
df_reduced.to_csv("../data/hotel_booking_clean.csv", index=False)

In [123]:
df_reduced.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces 

In [124]:
display_categorical_data_type_as_unqiue(df_reduced)

hotel
['Resort Hotel' 'City Hotel']
arrival_date_month
['July' 'August' 'September' 'October' 'November' 'December' 'January'
 'February' 'March' 'April' 'May' 'June']
meal
['BB' 'FB' 'HB' 'SC' 'No meal']
country
['PRT' 'GBR' 'USA' 'ESP' 'IRL' 'FRA' 'ROU' 'NOR' 'OMN' 'ARG' 'POL' 'DEU'
 'BEL' 'CHE' 'CN' 'GRC' 'ITA' 'NLD' 'DNK' 'RUS' 'SWE' 'AUS' 'EST' 'CZE'
 'BRA' 'FIN' 'MOZ' 'BWA' 'LUX' 'SVN' 'ALB' 'IND' 'CHN' 'MEX' 'MAR' 'UKR'
 'SMR' 'LVA' 'PRI' 'SRB' 'CHL' 'AUT' 'BLR' 'LTU' 'TUR' 'ZAF' 'AGO' 'ISR'
 'CYM' 'ZMB' 'CPV' 'ZWE' 'DZA' 'KOR' 'CRI' 'HUN' 'ARE' 'TUN' 'JAM' 'HRV'
 'HKG' 'IRN' 'GEO' 'AND' 'GIB' 'URY' 'JEY' 'CAF' 'CYP' 'COL' 'GGY' 'KWT'
 'NGA' 'MDV' 'VEN' 'SVK' 'FJI' 'KAZ' 'PAK' 'IDN' 'LBN' 'PHL' 'SEN' 'SYC'
 'AZE' 'BHR' 'NZL' 'THA' 'DOM' 'MKD' 'MYS' 'ARM' 'JPN' 'LKA' 'CUB' 'CMR'
 'BIH' 'MUS' 'COM' 'SUR' 'UGA' 'BGR' 'CIV' 'JOR' 'SYR' 'SGP' 'BDI' 'SAU'
 'VNM' 'PLW' 'QAT' 'EGY' 'PER' 'MLT' 'MWI' 'ECU' 'MDG' 'ISL' 'UZB' 'NPL'
 'BHS' 'MAC' 'TGO' 'TWN' 'DJI' 'STP' 'KNA' 'ETH' 'IRQ' 'HN